registry/src/registry.rs

```diff
@@ -132,16 +197,49 @@ where
         Ok(self.db.get_validators_by_index(indices).await?)
     }
 
+    /// List all operators in the registry.
     pub(crate) async fn list_operators(&mut self) -> Result<Vec<Operator>, RegistryError> {
         self.sync.wait_for_sync().await;
         Ok(self.db.list_operators().await?)
     }
 
+    /// Get operators by signer.
     pub(crate) async fn get_operators_by_signer(
         &mut self,
         signers: &[Address],
     ) -> Result<Vec<Operator>, RegistryError> {
         self.sync.wait_for_sync().await;
         Ok(self.db.get_operators_by_signer(signers).await?)
     }
+
+    /// Get the active validators that will propose in the given epoch
+    /// that are also registered in the registry.
+    pub(crate) async fn get_lookahead(&mut self, epoch: u64) -> Result<Lookahead, RegistryError> {
+        // 1. fetch the proposer duties from the beacon node
+        let proposer_duties = self.beacon.get_lookahead(epoch, false).await?;
+        let proposer_pubkeys = proposer_duties
+            .iter()
+            .map(|d| BlsPublicKey::from_bytes(&d.public_key).expect("valid BLS pubkey"))
+            .collect::<Vec<_>>();
+
+        // 2. fetch the registry entries from the database
+        self.sync.wait_for_sync().await;
+        let registry_entries = self.db.get_validators_by_pubkey(&proposer_pubkeys).await?;
+
+        // 3. map registry entries to their proposal slot. example result:
+        //
+        // 10936976: { validator_pubkey: 0x1234, operator: 0x5678, gas_limit: 1000000, rpc_endpoint: https://rpc.example.com }
+        // 10936977: { validator_pubkey: 0x9214, operator: 0x5678, gas_limit: 1000000, rpc_endpoint: https://rpc.example.com }
+        // 10936978: { validator_pubkey: 0x1983, operator: 0x5678, gas_limit: 1000000, rpc_endpoint: https://rpc.example.com }
+        let mut lookahead = HashMap::new();
```

Comment:
Nit but can we create a type alias in primitives, to be used in openAPI spec?

```rust
type Lookahead = HashMap<u64, Entry>
```

Reply:
it's already like that, I guess I could use `Lookahead::new()` instead to make it more clear though

Reply:
done [116d8f1](https://github.com/chainbound/bolt-registry/pull/35/commits/116d8f1b2d409f9a26c66c1fcda0d53d85db1ce9)